File: src/wasm/hexchess/pieces/king.rs
use crate::constants::Color;
use crate::hexchess::hexchess::Hexchess;
use crate::hexchess::san::San;

use crate::hexchess::utils::{
    get_color,
    step,
};


pub fn king_moves_unsafe(
    hexchess: &Hexchess,
    from: u8,
    color: &Color,
) -> Vec<San> {
    let mut result: Vec<San> = vec![];

    let directions: [u8; 12] = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11];

    for n in directions {
        let to = match step(from, n) {
            Some(to) => to,
            None => continue,
        };

        match hexchess.board[to as usize] {
            Some(piece) => {
                if get_color(&piece) != *color {
                    result.push(San { from, promotion: None, to });
                }
            },
            None => {
                result.push(San { from, promotion: None, to });
            }
        };
    }

    result
}

File: php/src/Pieces/King.php
<?php

namespace Bedard\Hexchess\Pieces;

use Bedard\Hexchess\Board;
use Bedard\Hexchess\Hexchess;
use Bedard\Hexchess\San;

class King
{
    public static function moves(Hexchess $hexchess, int $from, string $color): array
    {
        $moves = [];

        for ($i = 0; $i < 12; $i++) {
            $to = Board::step($from, $i);

            if ($hexchess->board[$to] === null || Board::color($hexchess->board[$to]) !== $color) {
                $moves[] = new San($from, $to);
            }
        }

        return $moves;
    }
}
